### ISSUE TITLE
TabulatedFunction and InteractionGroups complie wrong on CUDA and OpenCL platforms

### ISSUE THREAD

**Issue by Lnaden** [pandegroup/openmm#1441]
If one tries to use `Continuous1DFunction` and `InteractionGroup`s in `CustomNonbondedForce` on the OpenCL and CUDA platforms, the following errors are thrown:

* OpenCL 
```python
  File "/home/ln8dc/anaconda/lib/python2.7/site-packages/simtk/openmm/openmm.py", line 14180, in __init__
    this = _openmm.new_Context(*args)
Exception: Error compiling kernel: <kernel>:146:16: error: use of undeclared identifier 'table0'
float4 coeff = table0[index];
```

* CUDA
```python
  File "/home/ln8dc/anaconda/lib/python2.7/site-packages/simtk/openmm/openmm.py", line 14180, in __init__
    this = _openmm.new_Context(*args)
Exception: Error compiling program: (726): error: identifier "table0" is undefined
(709): warning: variable "r" was declared but never referenced
1 error detected in this compilation.
```

I think the errors stem from the respective OpenCLExpressionUtilities.cpp and CudaExpressionUtilities.cpp in handling `TabulatedFunctions`. No error is raised on CPU or Reference platforms.

Here is minimal test case + particle positions, I have not tested more than `Continuous1DFunction`
```python
import simtk.openmm as mm
import simtk.unit as unit
import numpy as np

nparticle = 1
x= np.linspace(0,1, 11)
tf1d = mm.Continuous1DFunction(x, 0, 1)
pos = np.zeros([nparticle,3])
s = mm.System()

expression = "myfunc(myVar);"
cnbf = mm.CustomNonbondedForce(expression)
cnbf.addTabulatedFunction("myfunc", tf1d)
cnbf.addGlobalParameter('myVar',1)
cnbf.addInteractionGroup(set([0]), set([0]))

for i in xrange(nparticle):
    s.addParticle(12)
    pos[i,0] = i
    cnbf.addParticle([])
s.addForce(cnbf)

integrator = mm.VerletIntegrator(1*unit.femtosecond)
platform = mm.Platform.getPlatformByName('OpenCL')
context = mm.Context(s, integrator, platform) #Point of Failure
context.setPositions(pos)
print context.getState(getEnergy=True).getPotentialEnergy()
```

**Comment by peastman** [comment]
I just sent a PR with the fix.  Thanks for the excellent test case!

**Comment by Lnaden** [comment]
I'd like to reopen this issue as I found a subtle but related bug.

It appears that changing `TabulatedFunction`s on `Force`s with interaction groups, then calling `context.reinitialize()` does not actually change the computed `TabulatedFunction`, despite being appearing to change. Without interaction groups, the `TabulatedFunction` is updated correctly. Issue only occurs on OpenCL and CUDA platforms, Reference and CPU platforms update correctly.

Here is the modified test case to replicate the error:
```python
import simtk.openmm as mm
import simtk.unit as unit
import numpy as np
from copy import deepcopy

xmin = 0
xmax = 1
xcount = 3
nparticle = 2
x= np.linspace(xmin,xmax, xcount)
x2 = np.linspace(xmin,xmax,xcount)**2 #Test tabulated function update
tf1d = mm.Continuous1DFunction(x, xmin, xmax)
pos = np.zeros([nparticle,3])

s = mm.System()
expression = "myfunc(myVar);"
cnbf = mm.CustomNonbondedForce(expression)
cnbf.addTabulatedFunction("myfunc", tf1d)
cnbf.addGlobalParameter('myVar',0.5)
cnbf2 = deepcopy(cnbf)
set0 = set([0])
set1 = set([1])
cnbf.addInteractionGroup(set0, set1)

for i in xrange(nparticle):
    s.addParticle(12)
    pos[i,0] = i
    cnbf.addParticle([])
    cnbf2.addParticle([])
s2 = deepcopy(s)
s.addForce(cnbf)
s2.addForce(cnbf2)

ts = 2*unit.femtosecond
integrator = mm.VerletIntegrator(ts)
integrator2 = mm.VerletIntegrator(ts)
platform = mm.Platform.getPlatformByName('OpenCL')

contextWithIGroups = mm.Context(s, integrator, platform)
contextNoIGroups = mm.Context(s2, integrator2, platform)
for (name, context, force) in zip(['With Interaction Groups', 'No Interaction Groups'], [contextWithIGroups, contextNoIGroups], (cnbf, cnbf2)):
    context.setPositions(pos)
    state = context.getState(getEnergy=True, getPositions=True)
    print name
    print "Potential energy should be 0.5"
    print state.getPotentialEnergy()
    print context.getSystem().getForce(0).getFunctionParameters(0)
    #Update tabulated function
    force.setFunctionParameters(0, 'myfunc', x2, xmin, xmax)
    context.reinitialize()
    context.setPositions(pos)
    #Generate new energy, show function has been updated
    state = context.getState(getEnergy=True, getPositions=True)
    print "Potential energy should be 0.25"
    print state.getPotentialEnergy()
    print context.getSystem().getForce(0).getFunctionParameters(0)
    print("\n")
```
Returns the following:
```python
With Interaction Groups
Potential energy should be 0.5
0.5 kJ/mol
['myfunc', (0.0, 0.5, 1.0), 0.0, 1.0]
Potential energy should be 0.25
0.5 kJ/mol
['myfunc', (0.0, 0.25, 1.0), 0.0, 1.0]

No Interaction Groups
Potential energy should be 0.5
0.5 kJ/mol
['myfunc', (0.0, 0.5, 1.0), 0.0, 1.0]
Potential energy should be 0.25
0.25 kJ/mol
['myfunc', (0.0, 0.25, 1.0), 0.0, 1.0]
```

**Comment by Lnaden** [comment]
I also cannot directly re-open this issue as @peastman closed it.

**Comment by peastman** [comment]
I doubt it's actually a problem with `reinitialize()`.  That completely throws away the ContextImpl and creates a new one from scratch, so it would be *really* hard for any information to get kept around.  I'm more suspicious of something in the Python layer.  Could you try reversing how you create the objects?  That is, add the interaction group to `cnbf2` instead of `cnbf`.  In that case, which one produces the wrong result?

**Comment by Lnaden** [comment]
Adding the interaction groups to the `cnbf2` causes the `Context` its assigned to have the problem instead, even if the interaction group is added after `System` creation. I can also create the issue without doing any `copy` operations and just building one `CustomNonbondedForce` with the `TabulatedFunction`, one `System` and one `Context`. I had included the extra `Context` without the interaction group as a control.

**Comment by Lnaden** [comment]
Awesome! Thanks @peastman!

Fixed by #1444
